File: leetcode_cpp/90. Subsets II.cpp
// time complexity O(n^2)
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> path;
        sort(nums.begin(), nums.end());
        dfs(nums, 0, path, res);
        return res;
    }

    void dfs(vector<int>& nums, int index, vector<int>& path, vector<vector<int>>& res) {
        res.push_back(path);
        for (int i=index; i<nums.size();i++) {
            if (i > index && nums[i] == nums[i-1]) continue;
            path.push_back(nums[i]);
            dfs(nums, i+1, path, res);
            path.pop_back();
        }
    }
};

File: leetcode/90. Subsets II.py
# time complexity O(n^2)
class Solution:
    def subsetsWithDup(self, nums):
        res = []
        self.dfs(sorted(nums), 0, [], res)
        return res
    
    def dfs(self, nums, index, path, res):
        res.append(path)
        for i in range(index, len(nums)):
            if i > index and nums[i] == nums[i-1]:  #當層看 ex:[1,2,2], 確保i = index時 該element 依然被計算, i > index時 且nums[i] == nums[i-1] 則跳過
                continue
            self.dfs(nums, i+1, path+[nums[i]], res)

#刷題用這個
class Solution:
    def subsetsWithDup(self, nums):
        res = []
        self.dfs(sorted(nums), 0, [], res)
        return res
    
    def dfs(self, nums, index, path, res):
        res.append(path.copy())
        for i in range(index, len(nums)):
            if i > index and nums[i] == nums[i-1]:  #當層看 ex:[1,2,2], 確保i = index時 該element 依然被計算, i > index時 且nums[i] == nums[i-1] 則跳過
                continue
            path.append(nums[i])
            self.dfs(nums, i+1, path, res)
            path.pop()



''' 
Given a collection of integers that might contain duplicates, nums, return all possible subsets (the power set).

Note: The solution set must not contain duplicate subsets.

Example:

Input: [1,2,2]
Output:
[
  [2],
  [1],
  [1,2,2],
  [2,2],
  [1,2],
  []
]


'''